Breakout Magnetometers

shihaocao opened the pull request:
# Breakout Magnetometers
The objective of this PR is to be able to interact with MAG1 and MAG2 seperately, and simultaneously query both MAG's over I2C.

Addresses #212 
### Summary of changes
- Update constants.hpp (adcs), many registers/constants were renamed
- Update ADCSSW to support magnetometer breakout
- Updated ADCS Driver for ^
- Updated FSW Control Tasks for ^
- Updated FSW Unit tests for ^
- Updated ADCSCheckoutCase
- Updated flow_data.csv
- Minor QOL /tools/ changes

### Testing
Ptest testing shows that the whole loop is capable of reading from both MAG's simultaneously. This is sufficient for this PR.
Further nuanced ADCS ptesting will come with upgrades to the ADCSCheckoutCase (added to ticket).

### Constants
No real changes.

### Telemetry
Magnetometer related state fields were both added and renamed.
Telemetry was added to `flow_data.csv`

### Documentation Evidence
Added to ReadTheDocs

shihaocao commented on src/fsw/FCCode/AttitudeEstimator.cpp:
```diff
@@ -43,8 +44,11 @@ void AttitudeEstimator::set_data(){
     const d_vector_t r_ecef = pos_vec_ecef_fp->get();
     data.r_ecef = {r_ecef[0], r_ecef[1], r_ecef[2]};
 
```
@kkrol27 How would you like to handle the section below?

shihaocao commented on src/fsw/FCCode/AttitudeEstimator.cpp:
```diff
@@ -43,8 +44,11 @@ void AttitudeEstimator::set_data(){
 
-    const f_vector_t mag_vec = mag_vec_fp->get();
-    data.b_body = {mag_vec[0], mag_vec[1], mag_vec[2]};
+    const f_vector_t mag1_vec = mag1_vec_fp->get();
```
@kkrol27 How'd you like to handle the section below?

kkrol27 replied:
Honestly, for the time being we can just pull magnetometer one or something. With my attitude controller and estimator rewrite I can set up some logic to handle it then.

shihaocao replied:
Yep sounds good

kkrol27 commented on src/fsw/FCCode/ADCSBoxMonitor.cpp:
```diff
@@ -124,7 +129,8 @@ void ADCSBoxMonitor::execute(){
     ssa_voltages.fill(0);
 
-    f_vector_t mag_vec;
+    f_vector_t mag1_vec;
+    f_vector_t mag2_vec;
```
If you're going through this atm, would you mind switching things over to `lin::Vector3f` state fields instead? I know most of the flight computer is going in that direction and the `AttitudeEstimator` will sort of expect it.

shihaocao replied:
Done

shihaocao commented:
I just now added a new section to ADCSCheckoutCase to make sure that magnetometers could function independently, caught a bug, just removed some `return` lines in `imu.cpp`

kkrol27 commented on src/fsw/FCCode/ADCSBoxMonitor.cpp:
```diff
@@ -115,6 +115,14 @@ bool exceed_bounds(const float input, const float min, const float max){
 }
 
+// TODO: WOULD BE A NICE FEATURE IN LIN.hpp
+template<size_t N>
+lin::Vector<float, N> to_linvector(const std::array<float, N>& src) {
+    lin::Vector<float, N> src_cpy;
+    for(unsigned int i = 0; i < 3; i++) src_cpy(i) = src[i];
+    return src_cpy;
+}
```
This should be marked `static` and the range in the for loop needs to be based on `N`.

shihaocao commented:
Yep, I re-ran HITL on these new changes. @kkrol27

shihaocao commented on constants:
```diff
@@ -3,35 +3,33 @@ Keep this file in version control so that changes to constants are well-known.
-src/adcs/constants.hpp:189: "max_voltage_thresh" = "max_voltage_rd"
-src/adcs/constants.hpp:191: "num_sun_sensors" = "20"
-src/adcs/constants.hpp:198: "max_devices" = "32"
+src/adcs/constants.hpp:94: "max_eq_temp" = "85.0f"
+src/adcs/constants.hpp:97: "min_eq_temp" = "-40.0f"
+src/adcs/constants.hpp:101: "max_rd_temp" = "25.0f + 128.0f"
+src/adcs/constants.hpp:103: "max_rd_temp" = "25.0f + 128.0f"
+src/adcs/constants.hpp:110: "min_rd_temp" = "25.0f - 128.0f"
+src/adcs/constants.hpp:112: "min_rd_temp" = "25.0f - 128.0f"
+src/adcs/constants.hpp:118: "max_rd_omega" = "125.0f * 0.01745329251f"
+src/adcs/constants.hpp:121: "min_rd_omega" = "-max_rd_omega"
+src/adcs/constants.hpp:125: "max_mag1_rd_mag" = "0.0050f"
+src/adcs/constants.hpp:128: "min_mag1_rd_mag" = "-max_mag1_rd_mag"
+src/adcs/constants.hpp:132: "max_mag2_rd_mag" = "0.0032f"
+src/adcs/constants.hpp:135: "min_mag2_rd_mag" = "-max_mag2_rd_mag"
+src/adcs/constants.hpp:142: "max_moment" = "0.113337f / 2.0f"
+src/adcs/constants.hpp:144: "min_moment" = "-max_moment"
+src/adcs/constants.hpp:154: "moment_of_inertia" = "0.0000135f"
+src/adcs/constants.hpp:156: "max_torque" = "310.1852f * moment_of_inertia"
+src/adcs/constants.hpp:158: "min_torque" = "-max_torque"
+src/adcs/constants.hpp:160: "max_speed_read" = "1047.20f"
+src/adcs/constants.hpp:162: "min_speed_read" = "-max_speed_read"
+src/adcs/constants.hpp:165: "max_speed_command" = "680.678f"
+src/adcs/constants.hpp:167: "min_speed_command" = "-max_speed_command"
+src/adcs/constants.hpp:174: "min_voltage_rd" = "0.0f"
+src/adcs/constants.hpp:176: "max_voltage_rd" = "3.3f"
+src/adcs/constants.hpp:178: "min_voltage_thresh" = "min_voltage_rd"
+src/adcs/constants.hpp:180: "max_voltage_thresh" = "max_voltage_rd"
+src/adcs/constants.hpp:182: "num_sun_sensors" = "20"
+src/adcs/constants.hpp:189: "max_devices" = "32"
```
No numerical changes were made, just two faulty lines were removed + line numbers changed.
